Split Ramen into 2 controllers, hub and dr-cluster, and create olm bundles for the them

ShyamsundarR opened the pull request:
PR contains 3 commits (with some overlaps):
- Split Ramen into 2 controllers, DRManager and DROrchestrator
- Change drmanager/orchestrator names to hub/dr-cluster
- Create bundle builds for hub and dr-cluster

Changes to deployments:
- There are now 2 deployments when `make deploy` is run, `ramen-dr-cluster-operator` and `ramen-hub-operator`
- If running outside the cluster the make target is changed from `make run` to `make run-hub` or `make run-dr-cluster` as required

### When to build bundles:
- Changes to CRD (new/updates)
  - Changes to samples (this should really be done when CRD changes, so should be covered by that item)
- Changes to roles
- Changes to deployment
- Changes to RamenConfig
- Changes to CSV

### When to update config/[dr-cluster|hub]/manifests/bases/*clusterserviceversion* file (the CSV):
- We are not using `operator-sdk generate kustomize manifests -q` hence the csv file needs to be updated by contributors
- Updated when a new CRD is added

### How to build and push a bundle:
- Example here is to use the non-ramen official quay registry for testing:
  - The bundle pulls an image to deploy, so if you need a specific image build and push that first
    - `IMG=quay.io/shyamsundarr/ramen-operator:canary make docker-build`
    - `IMG=quay.io/shyamsundarr/ramen-operator:canary make docker-push`
  - Build and push the bundle:
    - `IMAGE_TAG_BASE=quay.io/shyamsundarr/ramen IMG=quay.io/shyamsundarr/ramen-operator:canary make bundle-build bundle-push`
    - NOTE: For podman systems, do not use the `bundle-push` target, instead use `podman push` directly

### Deploying a built bundle:
- Bundle deployment requires olm installed in the cluster, so install olm (one-time operation),
  - `operator-sdk olm install`
- Roll out a dev version of the bundle,
  - Create the namespace for the bundle: `kubectl create ns ramen-system`
  - `operator-sdk run bundle quay.io/shyamsundarr/ramen-hub-operator-bundle:v0.0.1 -n ramen-system`
  - `operator-sdk run bundle quay.io/shyamsundarr/ramen-dr-cluster-operator-bundle:v0.0.1 -n ramen-system`
  - NOTE: Bundle needs to be available in a registry and not a cached image on the cluster

### Cleanup an installed bundle:
- `operator-sdk cleanup ramen-dr-cluster`
- `operator-sdk cleanup ramen-hub`

### TODOs:
- A catalog for the bundle is not yet available, hence bundles have to be installed using operator-sdk CLI
- `make undeploy` deletes the `ramen-system` namespace, so selectively undeploying hub|dr-cluster is only possible if they are installed in different namespaces
- Bundles currently have no dependencies, which need to be added
- Find ways to avoid "--load_restrictor none" in kustomize
- Documentation for this and related bundle installs
- github actions to auto-publish bundles and also check if bundles related files need updates

### Gotchas:
- Sample RamenConfig cannot be in a bundle
  - It errors out when running `operator-sdk generate bundle` with the error `accumulateFile error: \"accumulating resources from '../../samples/ramendr_v1alpha1_ramenconfig.yaml': missing metadata.name in object`

Signed-off-by: Shyamsundar Ranganathan <srangana@redhat.com>

ShyamsundarR commented:
@umangachapagain PTAL from a future split on same/similar lines to creating separate bundles as well.

vdeenadh commented on main.go:
```diff
@@ -65,13 +65,18 @@ func newManager() (ctrl.Manager, error) {
+	flag.StringVar(&leaderElectionID, "leader-election-id", "ae40cfcb.openshift.io",
+		"The leader election ID used by the controller manager if leader election is enabled.")
+	flag.StringVar(&controllerType, "controller-type", "invalid",
+		"The type of controller to execute by the controller manager.")
```
I said elsewhere that this flag should be moved to RamenConfig.  However, for ease of `make run`, I believe we can override RamenConfig with a command line flag.

ShyamsundarR replied:
`make run` can take a `--config=...` option, so I would look at this as a config only option even in the future.

vdeenadh reviewed (commented):
If we want a to run both drmanager and drorchestrator in the same cluster, this PR will require deploying two separate pods, right?  Did you consider using the controller-type flag to either start dr-manager reconciler or dr-orchestrator reconciler or both in the same ramen pod?

ShyamsundarR commented:
> If we want a to run both drmanager and drorchestrator in the same cluster, this PR will require deploying two separate pods, right? Did you consider using the controller-type flag to either start dr-manager reconciler or dr-orchestrator reconciler or both in the same ramen pod?

yes, I did consider this, but as we need different dependent CRs and such dropped the idea to run both within the same deployment.

ShyamsundarR commented on config-dro/default/kustomization.yaml:
```diff
@@ -0,0 +1,91 @@
+        - "--metrics-bind-address=127.0.0.1:8080"
+        - "--leader-elect"
+        - "--leader-election-id=drorchestrator.openshift.io"
+        - "--controller-type=drorchestrator"
```
Ack! this is in the interim till we get #93 merged, at which point we would move to the config instead.

ShyamsundarR commented on config-dro/default/kustomization.yaml:
```diff
@@ -0,0 +1,91 @@
+    group: apps
+    version: v1
+    kind: Deployment
+    name: controller-manager
```
Yes, the name is a bit too long. While at this, should we name/rename anything else?

BenamarMk replied:
Or maybe:
ramen-drmanager-controller
ramen-drorchestrator-controller

ShyamsundarR replied:
This is not trivial and there are multiple resources named as follows,
```
$ kubectl get clusterroles -n ramen-system | grep ramen
ramen-drm-manager-role                                                 2021-07-14T19:42:39Z
ramen-drm-metrics-reader                                               2021-07-14T19:42:39Z
ramen-drm-proxy-role                                                   2021-07-14T19:42:39Z
ramen-dro-manager-role                                                 2021-07-14T19:42:37Z
ramen-dro-metrics-reader                                               2021-07-14T19:42:37Z
ramen-dro-proxy-role                                                   2021-07-14T19:42:37Z

$ kubectl get clusterrolebinding -n ramen-system | grep ramen
ramen-drm-manager-rolebinding                          ClusterRole/ramen-drm-manager-role                                                 4m27s
ramen-drm-proxy-rolebinding                            ClusterRole/ramen-drm-proxy-role                                                   4m27s
ramen-dro-manager-rolebinding                          ClusterRole/ramen-dro-manager-role                                                 4m29s
ramen-dro-proxy-rolebinding                            ClusterRole/ramen-dro-proxy-role                                                   4m29s

$ kubectl get all -n ramen-system
NAME                                                   TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)    AGE
service/ramen-drm-controller-manager-metrics-service   ClusterIP   10.107.168.155   <none>        8443/TCP   6m54s
service/ramen-dro-controller-manager-metrics-service   ClusterIP   10.101.62.2      <none>        8443/TCP   6m56s

NAME                                           READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/ramen-drm-controller-manager   1/1     1            1           6m55s
deployment.apps/ramen-dro-controller-manager   1/1     1            1           6m56s
```

I really wanted to leverage the auto generated manifests, and not have to poke around there. For example this can be solved if we edit the manifests that are auto generated, but that may get overwritten each time we generate the same.

I can only rename the deployment and leave others as is (everything will get the name prefix of `ramen-drm` or the new `ramen-drmanager`). The deployments will be named as `ramen-drmanager-controller` and `ramen-drorchestrator-controller`. Is this acceptable, or should we leave it as is?

vdeenadh replied:
deployment.apps/ramen-drm-controller-manager 
deployment.apps/ramen-dro-controller-manager

1. One letter difference in a long string of words is hard to process. 
2. The manager suffix can also be confusing with our dr-manager.  

Instead of dr-manager, should we use something like kluster or mcluster?
deployment.apps/ramendr-hub-manager
deployment.apps/ramendr-kluster-manager

Also note the hyphenation of ramendr-hub makes the "hub" more readable than ramen-drhub

ShyamsundarR replied:
So only deployments need to be renamed? Leave the rest as is? Renaming everything would be laborious.

> deployment.apps/ramen-drm-controller-manager
> deployment.apps/ramen-dro-controller-manager
> 
>     1. One letter difference in a long string of words is hard to process.

Ack!

> 
>     2. The manager suffix can also be confusing with our dr-manager.

Absolutely

> 
> 
> Instead of dr-manager, should we use something like kluster or mcluster?
> deployment.apps/ramendr-hub-manager
> deployment.apps/ramendr-kluster-manager
> 
> Also note the hyphenation of ramendr-hub makes the "hub" more readable than ramen-drhub

Here is what I think (basically we are revisiting drmanager and drorchestrator notation, which is good):
**NOTE:** For ease of using auto generated semantics I prefer to retain the controller-manager suffix
ramen-hub-controller-manager
ramen-spoke-controller-manager

vdeenadh replied:
Got it.  I desperately tried to avoid suggesting `spoke` in my earlier comment because I think it is a misuse of the word spoke to refer to a managed cluster. Spoke is used to describe the path or edge between the hub and the nodes. But the alternative of `spoke nipple` (bicycle term) doesn't help either!  If you can't think of an alternative, spoke is fine with me; it matches with ACM's use of the term.

ShyamsundarR replied:
@BenamarMk or others, one last chance (for now) before I adapt to the hub/spoke names.... 😄

BenamarMk replied:
so hub and spoke I think sound good. But I think they are too heavy for ramen functionality.  
If we go with drm and dro, the drm will confuse people with digital right management that has been used in the past heavily. It is less used now adays though.  so I really don't want to go back to square 1, but I still think that if we drop manager from the end would be better.  And instead, it would be:
```
ramen-dro-controller
ramen-drmc-controller
```

BenamarMk reviewed (approved):
LGTM

umangachapagain reviewed (commented):
Split looks good to me.

ShyamsundarR commented:
@vdeenadh Updated main.go post rebasing to changes introduced by RamenConfig and also updated `make run` targets, PTAL.

I am working on renaming as per discussion [here](https://github.com/RamenDR/ramen/pull/111#discussion_r669268266) but current changes need a relook on the above topics, hence pinging for early comments.

ShyamsundarR commented:
@vdeenadh @BenamarMk Added a second commit that does the following,

```
Change drmanager/orchestrator names to hub/dr-cluster

Also changed,
- deployment name suffix from controller-manager to
operator
- Reflected above name change to other artifacts
- Removed unused auto generated kustomizations
- Changed ramen controller types to hub/drcluster
- Moved config files for hub/drcluster under config
- Changed config map name to not have ramen as a prefix

Kept as a separate commit to record the move from auto
generated artifacts to modified ones.
```

The change from controller-manager to operator is not conflicting with the `make generate|manifests` targets. So this should be future proof from changes in there.

Rules for auto-generated files update changes as follows:
- Changes to `config/rbac/role.yaml` needs to be reflected in the appropriate `config/hub/rbac/role.yaml` or `config/dr_cluster/role.yaml`
- Change to RamenConfig needs update in appropriate `config/[hub/dr_cluster]/manager/ramen_manager_config.yaml` AND in `examples/*config*`

vdeenadh commented on config/hub/default/kustomization.yaml:
```diff
@@ -6,7 +6,7 @@ namespace: ramen-system
 # Note that it should also match with the prefix (text before '-') of the namespace
 # field above.
-namePrefix: ramen-
+namePrefix: ramen-hub-
```
When we discussed, I believe we said only "hub".  But having seen a few places that use dr-hub, I like it equally, especially, it is consistent to say ramen-dr-hub-operator vs ramen-dr-cluster-operator.

ShyamsundarR replied:
I changed `drhub` in the leader election to `hub`, in an upcoming commit on this PR. To keep it consistent.

Name change across to call it drhub we can discuss further.

vdeenadh reviewed (approved):
> Change drmanager/orchestrator names to hub/dr-cluster
I believe you meant to say that you changed drmanager/orchestrator to dr-cluster/hub. Also, did you mean dr-hub or just hub? There are some places in the commit that uses dr-hub.

> - deployment name suffix from controller-manager to operator
Awesome

> - Changed config map name to not have ramen as a prefix
Should the config map names have dr-hub or dr-cluster prefix?

> The change from controller-manager to operator is not conflicting with the `make generate|manifests` targets. So this should be future proof from changes in there.
This is great.

> * Changes to `config/rbac/role.yaml` needs to be reflected in the appropriate `config/hub/rbac/role.yaml` or `config/dr_cluster/role.yaml`
> * Change to RamenConfig needs update in appropriate `config/[hub/dr_cluster]/manager/ramen_manager_config.yaml` AND in `examples/*config*`
Please add these reminders to the respective files as comments.

ShyamsundarR commented on config/hub/manager/kustomization.yaml:
```diff
@@ -0,0 +1,16 @@
+
+configMapGenerator:
+- files:
+  - ramen_manager_config.yaml
```
Leaving it as is, the filename is constant across hub and dr-cluster and the config map is created with the prefix.

ShyamsundarR commented on config/dr_cluster/manager/ramen_manager_config.yaml:
```diff
@@ -8,5 +8,5 @@ webhook:
   leaderElect: true
-  resourceName: leaderelection.ramendr.openshift.io
-#  resourceNamespace: ramen-system
+  resourceName: drcluster.ramendr.openshift.io
```
The config map is an artifact of leader election, so retaining the same format, i.e without the `leaderelection` prefix.

Changed this to `dr-cluster` in the leader election names across in the next commit to this PR.

ShyamsundarR commented on api/v1alpha1/ramenconfig_types.go:
```diff
@@ -25,6 +25,18 @@ import (
+
+const (
+	// DRCluster operates as the DR cluster controller on a peer cluster
+	DRCluster ControllerType = "dr-cluster"
```
I had similar thoughts on this, but left this as is. Change this field across in a subsequent PR? Or I can address this right away.

I also did not want to collide with any possible future `ControllerManagerConfigurationSpec` fields, and prefixed this with ramen for the json key only (jFYI), maybe confusing to read the field as `ControllerType` and JSON as `ramenControllerType` and so change it to one or the other?

ShyamsundarR commented on Makefile:
```diff
@@ -120,18 +123,41 @@ docker-push: ## Push docker image with the manager.
-deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in ~/.kube/config.
-	cd config/manager && $(KUSTOMIZE) edit set image controller=${IMG}
-	$(KUSTOMIZE) build config/default | kubectl apply -f -
+undeploy-hub: ## Undeploy DR Orchestrator controller from the K8s cluster specified in ~/.kube/config.
```
Addressed in the next commit on this PR.

ShyamsundarR commented on Makefile:
```diff
@@ -120,18 +123,41 @@ docker-push: ## Push docker image with the manager.
+uninstall-dr-cluster: manifests kustomize ## Uninstall DR Manager CRDs from the K8s cluster specified in ~/.kube/config.
+	$(KUSTOMIZE) build --load_restrictor none config/dr_cluster/crd | kubectl delete -f -
+
+deploy-dr-cluster: manifests kustomize ## Deploy DR Manager controller to the K8s cluster specified in ~/.kube/config.
```
Addressed in the next commit on this PR.

ShyamsundarR commented:
@vdeenadh @BenamarMk updated the PR description and also added a 3rd commit that brings in bundle building. This is an added commit, hence if prior commits are reviewed then you can just review the last one (just a heads up).

BenamarMk reviewed (approved):
LGTM
